src/net.cpp

```diff
@@ -2144,31 +2198,34 @@ void ThreadMessageHandler()
             if (pnode->fDisconnect)
                 continue;
 
-            // Receive messages from the net layer and put them into the receive queue.
-            if (!g_signals.ProcessMessages(pnode))
-                pnode->fDisconnect = true;
-
-            // Discover if there's more work to be done
-            if (pnode->nSendSize < SendBufferSize())
+            if (pnode->successfullyConnected())
             {
-                { // If already locked some other thread is working on it, so no work for this thread
-                    TRY_LOCK(pnode->csRecvGetData, lockRecv);
-                    if (lockRecv && (!pnode->vRecvGetData.empty()))
-                        fSleep = false;
-                }
-                if (fSleep)
-                { // If already locked some other thread is working on it, so no work for this thread
-                    TRY_LOCK(pnode->cs_vRecvMsg, lockRecv);
-                    if (lockRecv && (!pnode->vRecvMsg.empty() && pnode->vRecvMsg[0].complete()))
-                        fSleep = false;
-                }
+                // parallel processing
+                fSleep &= threadProcessMessages(pnode);
+            }
+            else
+            {
+                // serial processing during setup
+                TRY_LOCK(pnode->csSerialPhase, lockSerial);
```

Comment:
ok I see that you have serialized processing during setup (and I like the idea because it makes things easier to debug), so my prior 2 comments are not strictly necessary.  However, there seems to be no reason not to have them, especially setting state_incoming before sending VERACK.